pkg/v1/layout/write.go

```diff
@@ -0,0 +1,318 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package layout
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/empty"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+	"golang.org/x/sync/errgroup"
+)
+
+var layoutFile = `{
+    "imageLayoutVersion": "1.0.0"
+}`
+
+// AppendImage writes a v1.Image to an OCI image layout at path and updates
+// the index.json to reference it.
+func AppendImage(path string, img v1.Image, options ...LayoutOption) (v1.ImageIndex, error) {
+	if err := WriteImage(path, img); err != nil {
+		return nil, err
+	}
+
+	mt, err := img.MediaType()
+	if err != nil {
+		return nil, err
+	}
+
+	d, err := img.Digest()
+	if err != nil {
+		return nil, err
+	}
+
+	manifest, err := img.RawManifest()
+	if err != nil {
+		return nil, err
+	}
+
+	desc := v1.Descriptor{
+		MediaType: mt,
+		Size:      int64(len(manifest)),
+		Digest:    d,
+	}
+
+	for _, opt := range options {
+		if err := opt(&desc); err != nil {
+			return nil, err
+		}
+	}
+
+	return AppendDescriptor(path, desc)
+}
+
+// AppendIndex writes a v1.ImageIndex to an OCI image layout at path and updates
+// the index.json to reference it.
+func AppendIndex(path string, ii v1.ImageIndex, options ...LayoutOption) (v1.ImageIndex, error) {
+	if err := WriteIndex(path, ii); err != nil {
+		return nil, err
+	}
+
+	mt, err := ii.MediaType()
+	if err != nil {
+		return nil, err
+	}
+
+	d, err := ii.Digest()
+	if err != nil {
+		return nil, err
+	}
+
+	manifest, err := ii.RawIndexManifest()
+	if err != nil {
+		return nil, err
+	}
+
+	desc := v1.Descriptor{
+		MediaType: mt,
+		Size:      int64(len(manifest)),
+		Digest:    d,
+	}
+
+	for _, opt := range options {
+		if err := opt(&desc); err != nil {
+			return nil, err
+		}
+	}
+
+	return AppendDescriptor(path, desc)
+}
+
+// AppendDescriptor adds a descriptor to the index.json of an ImageIndex located
```

Comment:
I think this doc comment could be improved by talking in terms of an image layout. That might be simplest by turning it into a method as I mention below. But even if you decide against making this a method, I think it would help to raise the level of the description a bit. I know that an image layout is represented as an `index.json` of an `ImageIndex`, but calling it an image layout makes the intended usage clearer.